### ISSUE TITLE
IMAP/IMAP: UIDMaps file in incoherent state (truncated) regarding past synced UIDs

### ISSUE THREAD

**Issue by mobamoba** [OfflineIMAP/offlineimap#380]
#### General informations

- Ubuntu 16.04: 
- offlineimap version 7.07
- Python version: 2.7
- server name or domain: 
- CLI options: 

#### Configuration file offlineimaprc

```
[general]
accounts = 1, 2, 3, 4, 5, 6
maxsyncaccounts = 7
ignore-readonly = no
socktimeout = 60

[Account Exchange]
localrepository = Local
remoterepository = Remote
autorefresh = -1
maxconnections = 2
status_backend = sqlite

[Repository Local]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com
nametrans = lambda sentname: re.sub ('Sent Items', 'Sent Current',
				sentname)

[Repository Remote]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
nametrans = lambda sentname: re.sub ('Sent Current', 'Sent Items',
				sentname)
folderfilter = lambda foldername: foldername in [
	'INBOX', 'Saved Current', 'Sent Current', 'Drafts', 'Trash', '[Gmail]/Spam',
	'SMS', 'SMSArchived', 'Call log', 'Call log Archived']
keepalive = 60

[Account 1]
localrepository = Local-1
remoterepository = Remote-1
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-1]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-1]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
	'SLISent']
nametrans = lambda sentname: re.sub ('SLISent', 'Sent SLI',
				sentname)
readonly = True

[Account 2]
localrepository = Local-2
remoterepository = Remote-2
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-2]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-2]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
	'2Sent']
nametrans = lambda sentname: re.sub ('2Sent', 'Sent 2',
				sentname)
readonly = True

[Account 3]
localrepository = Local-3
remoterepository = Remote-3
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-3]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-3]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
	'3Sent']
nametrans = lambda sentname: re.sub ('3Sent', 'Sent 3',
				sentname)
readonly = True

[Account 4]
localrepository = Local-4
remoterepository = Remote-4
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-4]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-4]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
	'4Sent']
nametrans = lambda sentname: re.sub ('4Sent', 'Sent 4',
				sentname)
readonly = True

[Account 5]
localrepository = Local-5
remoterepository = Remote-5
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-5]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-5]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
	'5Sent']
nametrans = lambda sentname: re.sub ('5Sent', 'Sent 5',
				sentname)
readonly = True```

#### pythonfile (if any)

```
REMOVE PRIVATE DATA.
```


#### Logs, error

```
REMOVE PRIVATE DATA.
```

#### Steps to reproduce the error

- 
- 

I've been using Offlineimap for years and I've never seen this before, but today it permanently deleted 3 emails. I have no idea how to debug this and because I watched it happen in real time, I've shutdown Offlineimap because I can't afford to have it delete my emails.

All I can tell you is this: Offlineimap thinks it's syncing to my Gmail just fine. There are no errors, no anything, just normal syncing. However, the Inbox in my Gmail and the Inbox on my server (Exchange/Outlook) are completely different, i.e. they aren't syncing at all. When a new email arrives in my Exchange Inbox, instead of being synced to Gmail, it's wiped out by Offlineimap.

I watched this happen: Offlineimap noticed an email in my Inbox, said it was copying local to remote and instead deleted it from local and copied nothing to remote. It watched it vanish from my Outlook Inbox just as Offlineimap was doing this.

Will a debug log work on a dry run? If so, I'd be willing to do one but, because data's being lost, I can't  run Offlineimap for real in order to do a log as I'm sure you can understand.

**Comment by nicolas33** [comment]
What are the account names? How do you sync both accounts?

Did offlineimap remove already synced emails from a **previous** run or everything happened in the same run?

**Comment by mobamoba** [comment]
I'm syncing an Exchange account to a Gmail account. I receive my mail through Exchange and Gmail is essentially my backup.

Offlineimap thinks the inbox of the accounts are synced they're not actually. My Gmail account Inbox looks totally different from my Exchange Inbox - yet offlineimap isn't throwing any errors. 

Offlineimap removes the mail as soon as it detects it, i.e. a mail arrives in my Exchange Inbox; offlineimap notices; offlineimap deletes it rather than syncing it; offlineimap continues as if no my Exchange and Gmail accounts are perfectly in sync even though they're not.

**Comment by nicolas33** [comment]
Could you run a session with IMAP debug log enabled, please?

**Comment by mobamoba** [comment]
Does debug log work with dry run enabled?

**Comment by nicolas33** [comment]
Yes, but I'm not sure this will help. You might create other accounts for the same remotes to make backups.

**Comment by mobamoba** [comment]
I've attached two logs. I sent myself a test email between the two log runs, so oi2debug.txt is prior to receiving a new email in my inbox and oi3debug.txt is after receiving that test email.

For both runs, my Gmail inbox shows 2 emails. For oi2debug.txt, there are 4 emails in my Exchange inbox; for oi3debug.txt, there are 5, one of which is unread. The 2 emails in my Gmail inbox are completely different from any of the emails in my Exchange mailbox.

Even though this is a dry-run, oi2debug.txt should still show the difference between my Gmail inbox (2 emails which have since been archived in Exchange) and my Exchange inbox (4 brand new emails). Oi3debug.txt should show all that plus the new fifth unread email in Exchange.


[oi2debug.txt](https://github.com/OfflineIMAP/offlineimap/files/518354/oi2debug.txt)
[oi3debug.txt](https://github.com/OfflineIMAP/offlineimap/files/518355/oi3debug.txt)

**Comment by nicolas33** [comment]
Thanks. What would actually help is a run failing to sync a new email. Once both sides diverged, we can't see anymore *why* they diverged.

**Comment by mobamoba** [comment]
oi3debug.txt that I attached early should be exactly that - wasn't it?

**Comment by nicolas33** [comment]
> oi3debug.txt that I attached early should be exactly that - wasn't it?

Yes, sorry.

This oi3debug.txt debug log is hard to work on (+40000 lines). Please, limit the logs to `-d imap`. Limiting to the failing folder "IMAP" with folderfilter would greatly help, too.

In the metadata folder, you'll find the mapping of the UIDs for both sides in the "UIDMapping" sub-folder. Please, provide the current mapping before the run. I wonder it's broken.

You seem to say it's a regression. Could you find a working release with your current mapping files? If so, it would be great to know the exact release introducing this issue.

**Comment by mobamoba** [comment]
Could you give me the exact debug command you want me to run with all the arguments (the folder is Inbox)? Otherwise I'll likely guess wrong.

I'll save the UIDMapping.

It's not a regression. I was using whatever version was current (6? 7?) and then upgraded to the latest release after noticing the problem and hoping the latest release would fix it. It didn't.

**Comment by nicolas33** [comment]
`/usr/local/bin/offlineimap --dry-run -o -a Exchange -d imap -l oi3debug.log` with the folderfilter configuration option set to INBOX only.

**Comment by mobamoba** [comment]
The UIDMappings were empty for Inbox - the local was an empty file and the remote didn't exist at all. Here is the debug log with just Inbox being filtered:

2016-10-19 09:38:50 INFO: OfflineImap 7.0.7 starting...
  Python: 2.7.12 Platform: linux2
  Args: /usr/local/bin/offlineimap --dry-run -o -a Exchange -d imap -l oi3debug.log
2016-10-19 09:38:50 INFO: Debug mode: Forcing to singlethreaded.
2016-10-19 09:38:50 DEBUG: Now debugging for imap: IMAP protocol debugging
2016-10-19 09:38:50 DEBUG: Now debugging for : Other offlineimap related sync messages
2016-10-19 09:38:50 DEBUG: [imap]: Using authentication mechanisms ['GSSAPI', 'XOAUTH2', 'CRAM-MD5', 'PLAIN', 'LOGIN']
2016-10-19 09:38:50 DEBUG: [imap]: Using authentication mechanisms ['GSSAPI', 'XOAUTH2', 'CRAM-MD5', 'PLAIN', 'LOGIN']
2016-10-19 09:38:50 INFO: *** Processing account Exchange
2016-10-19 09:38:50 INFO: Establishing connection to imap.gmail.com:993 (Remote)
2016-10-19 09:38:51 DEBUG: [imap]: Attempting XOAUTH2 authentication
2016-10-19 09:38:51 DEBUG: [imap]:   38:51.34 Account sync Exchange last 20 log messages:
2016-10-19 09:38:51 WARNING: XOAUTH2 authentication failed: AUTHENTICATE command error: BAD ['Client aborted AUTHENTICATE command. 71mb139712302qgt']. Data: BBFF2 AUTHENTICATE XOAUTH2

2016-10-19 09:38:51 DEBUG: [imap]: Attempting PLAIN authentication
2016-10-19 09:38:51 DEBUG: [imap]: __plainhandler: returning admin@mattprager.com(passwd hidden for log)
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Call log'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Call log Archived'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Drafts'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Saved'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Saved/Gmail Saved 2012'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Saved/Gmail Saved 2013'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Saved/Gmail Saved 2014'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Saved/Gmail Saved 2015'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Sent'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Sent/Gmail Sent 2012'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Sent/Gmail Sent 2013'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Sent/Gmail Sent 2014'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Gmail Sent/Gmail Sent 2015'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Junk E-mail'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Results Count'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'SMS'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'SMS Archived2'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'SMSArchived'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Saved Current'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'Sent Current'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out 'WhatsApp'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out '[Gmail]/Chats'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out '[Gmail]/Drafts'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out '[Gmail]/Sent Mail'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out '[Gmail]/Spam'[Remote] due to folderfilter
2016-10-19 09:38:52 DEBUG: []: Filtering out '[Gmail]/Trash'[Remote] due to folderfilter
2016-10-19 09:38:52 INFO: Establishing connection to www.jivemonkeyinc.com:993 (Local)
2016-10-19 09:38:52 DEBUG: [imap]: Attempting GSSAPI authentication
2016-10-19 09:38:52 DEBUG: [imap]: Attempting PLAIN authentication
2016-10-19 09:38:52 DEBUG: [imap]: __plainhandler: returning MattPrager@jivemonkeyinc.com(passwd hidden for log)
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'RSS Feeds' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sent Stuff' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Contacts' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Suggested Contacts' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'News Feed' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sent TOP' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Journal' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Spam' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Outbox' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Junk E-Mail' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sync Issues/Local Failures' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sync Issues' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Notes' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Deleted Items' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sync Issues/Conflicts' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Tasks' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sent SLI' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder '[Gmail]' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sync Issues/Server Failures' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sent JMI' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Sent Shopping' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Not creating folder 'Calendar' (repository 'Remote') as it would be filtered out on that repository.
2016-10-19 09:38:52 DEBUG: []: Copying folder structure from Remote to Local
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder '[Gmail]/Chats'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder '[Gmail]/Drafts'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder '[Gmail]/Sent Mail'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder '[Gmail]/Spam'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder '[Gmail]/Trash'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Call log'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Call log Archived'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Drafts'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Saved'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Saved/Gmail Saved 2012'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Saved/Gmail Saved 2013'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Saved/Gmail Saved 2014'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Saved/Gmail Saved 2015'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Sent'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Sent/Gmail Sent 2012'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Sent/Gmail Sent 2013'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Sent/Gmail Sent 2014'[Remote]
2016-10-19 09:38:52 DEBUG: []: Not syncing filtered folder 'Gmail Sent/Gmail Sent 2015'[Remote]
2016-10-19 09:38:52 INFO: Syncing INBOX: Gmail -> MappedIMAP
2016-10-19 09:38:52 DEBUG: Loading message list for MappedIMAP[INBOX]
2016-10-19 09:38:53 DEBUG: [imap]: calling imaplib2 fetch command: '1:*' (FLAGS UID INTERNALDATE)
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"14-Oct-2016 13:46:00 -0400"', 'FLAGS': '(\\Seen \\Recent)', 'UID': '12979'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"14-Oct-2016 16:41:40 -0400"', 'FLAGS': '(\\Seen \\Recent)', 'UID': '13021'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"18-Oct-2016 05:02:49 -0400"', 'FLAGS': '(\\Seen \\Recent)', 'UID': '13151'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"18-Oct-2016 10:25:45 -0400"', 'FLAGS': '(\\Seen \\Recent)', 'UID': '13169'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"18-Oct-2016 11:48:17 -0400"', 'FLAGS': '(\\Seen \\Recent)', 'UID': '13179'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"18-Oct-2016 18:35:55 -0400"', 'FLAGS': '(\\Seen \\Recent)', 'UID': '13219'}
2016-10-19 09:38:54 DEBUG: Message list for MappedIMAP[INBOX] loaded: 6 messages
2016-10-19 09:38:54 DEBUG: Loading message list for Gmail[INBOX]
2016-10-19 09:38:54 DEBUG: [imap]: calling imaplib2 fetch command: '1:*' (FLAGS UID INTERNALDATE)
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"27-Sep-2016 22:18:42 +0000"', 'FLAGS': '(\\Seen)', 'UID': '44901'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"25-Sep-2016 22:30:44 +0000"', 'FLAGS': '(\\Seen)', 'UID': '45055'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"14-Oct-2016 21:01:24 +0000"', 'FLAGS': '()', 'UID': '45056'}
2016-10-19 09:38:54 DEBUG: [imap]: __options2hash returning: {'INTERNALDATE': '"18-Oct-2016 04:02:40 +0000"', 'FLAGS': '()', 'UID': '45057'}
2016-10-19 09:38:54 DEBUG: Message list for Gmail[INBOX] loaded: 4 messages
2016-10-19 09:38:54 DEBUG: Syncing messages Gmail[INBOX] -> MappedIMAP[INBOX]
2016-10-19 09:38:54 INFO: [DRYRUN] Copy 2 messages from INBOX[Remote] to Local
2016-10-19 09:38:54 DEBUG: Syncing messages MappedIMAP[INBOX] -> Gmail[INBOX]
2016-10-19 09:38:54 INFO: [DRYRUN] Copy 6 messages from INBOX[Local] to Remote
2016-10-19 09:38:54 INFO: [DRYRUN] Deleting 2 messages (44901,45055) in Gmail[INBOX]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'Junk E-mail'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'Results Count'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'Saved Current'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'Sent Current'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'SMS'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'SMS Archived2'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'SMSArchived'[Remote]
2016-10-19 09:38:54 DEBUG: []: Not syncing filtered folder 'WhatsApp'[Remote]
2016-10-19 09:38:55 INFO: *** Finished account 'Exchange' in 0:04

**Comment by nicolas33** [comment]
> The UIDMappings were empty for Inbox

I don't get why unless there's no synced emails for the INBOX folder.

> Here is the debug log with just Inbox being filtered

I wonder the --dryn-run CLI option is not showing what would actually be done. I see no reason why it pretends "Deleting 2 messages (44901,45055) in Gmail[INBOX]". At first, I was thinking of a broken mapping for "conflicting" UIDs from both sides but the ranges are not even overlapping.

What's the output of the following command?
`echo 'SELECT * FROM status;' | sqlite3 <metadata>/Account-Exchange/Localstatus-sqlite/INBOX`

**Comment by mobamoba** [comment]
It says sqlite3 isn't installed. Do you want me to install it?

**Comment by nicolas33** [comment]
Yes, please.

**Comment by mobamoba** [comment]
44901|S|0|                                                                                                                           
45055|S|0|

**Comment by nicolas33** [comment]
Here we are. Make a backup of `<metadata>/Account-Exchange/Localstatus-sqlite/INBOX` and remove this file. Then, you should be able to sync again.

It appears that offlineimap had already synced both emails. For some reason, the mapping file is not coherent (truncated) so that it can't map those UIDs from one side to the other. Since offlineimap sees that the 44901 and 45055 UIDs have been synced in the past and cannot map them anymore to check if they still exist, it considers the deletions must be propagated.

My best guess is that offlineimap was shut down unexpectedly leaving a truncated mapping file. I expect this issue to be a very rare bug. I'll check the code.

**Comment by mobamoba** [comment]
That seems to have solved the problem - thanks. And, yes, would be great if OI would throw an error in this situation rather than delete since when OI deletes it's permanent (rather than going to trash).

**Comment by nicolas33** [comment]
Could you test above patch to check this won't hurt a normal run, please?

**Comment by mobamoba** [comment]
Sure. Should I run OI as a dry run after installing or run normal?

**Comment by nicolas33** [comment]
Both.

**Comment by mobamoba** [comment]
I'll do dry run first then post my result. If you don't see any data loss from that, I'll try it in a regular run after that.

**Comment by nicolas33** [comment]
This made me think I was not plain sure the map file was untouched in dry run mode. I've added a patch to the branch to make this certain.

**Comment by mobamoba** [comment]
Here are the results of the dry run - do you see any errors or should I go ahead and run it in normal mode?


2016-10-26 07:07:14 INFO: OfflineImap 7.0.8 starting...
  Python: 2.7.12 Platform: linux2
  Args: /usr/local/bin/offlineimap --dry-run -o -a Exchange -d imap -l oi6debug.log
2016-10-26 07:07:14 INFO: Debug mode: Forcing to singlethreaded.
2016-10-26 07:07:14 DEBUG: Now debugging for imap: IMAP protocol debugging
2016-10-26 07:07:14 DEBUG: Now debugging for : Other offlineimap related sync messages
2016-10-26 07:07:14 DEBUG: [imap]: Using authentication mechanisms ['GSSAPI', 'XOAUTH2', 'CRAM-MD5', 'PLAIN', 'LOGIN']
2016-10-26 07:07:14 DEBUG: [imap]: Using authentication mechanisms ['GSSAPI', 'XOAUTH2', 'CRAM-MD5', 'PLAIN', 'LOGIN']
2016-10-26 07:07:14 INFO: *** Processing account Exchange
2016-10-26 07:07:14 INFO: Establishing connection to imap.gmail.com:993 (Remote)
2016-10-26 07:07:15 DEBUG: [imap]: Attempting XOAUTH2 authentication
2016-10-26 07:07:15 DEBUG: [imap]:   07:15.36 Account sync Exchange last 20 log messages:
2016-10-26 07:07:15 WARNING: XOAUTH2 authentication failed: AUTHENTICATE command error: BAD ['Client aborted AUTHENTICATE command. g59mb237281631qva']. Data: MJJI2 AUTHENTICATE XOAUTH2

2016-10-26 07:07:15 DEBUG: [imap]: Attempting PLAIN authentication
2016-10-26 07:07:15 DEBUG: [imap]: __plainhandler: returning

**Comment by nicolas33** [comment]
I see no errors. Please, go ahead normal for some time.

**Comment by mobamoba** [comment]
Okay running it normal on all my accounts. Haven't seen anything bad so far. When in the process does OI commit the data from the tmp file to the actual map file?

**Comment by nicolas33** [comment]
Sure: https://github.com/OfflineIMAP/offlineimap/commit/2ac68d82aaa2a3651662888cb3e9aff21c6f92a2#diff-4470289ded51f44c1f37013bb78a2206R115

**Comment by mobamoba** [comment]
I see. I run OI for 6 accounts with autorefresh set to -1 and I've noticed that OI doesn't really respond to USR1, USR2, or even a terminate signal, so I'm guessing at some point I killed OI because I was unable to get it to exit gracefully and that's why this mapping problem occurred. Sounds like the solution you implemented would take care of that and I've been running it on all the accounts for the past few hours and haven't seen any errors.
